pest/src/stack.rs

```diff
@@ -0,0 +1,177 @@
+// pest. The Elegant Parser
+// Copyright (c) 2018 Dragoș Tiselice
+//
+// Licensed under the Apache License, Version 2.0
+// <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0> or the MIT
+// license <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. All files in the project carrying such notice may not be copied,
+// modified, or distributed except according to those terms.
+
+/// Implementation of a `Stack` which maintains an log of `StackOp`s in order to rewind the stack
+/// to a previous state.
+#[derive(Debug)]
+pub struct Stack<T: Clone> {
+    ops: Vec<StackOp<T>>,
+    cache: Vec<T>,
+    snapshots: Vec<usize>
```

Comment:
It might be a bit early to say this, but `snapshots` may end up very similar to the `queue.truncate`-ion that we use in `ParserState`. In case it does, it makes sense to have one way to do this. (either on the call stack or in a `Vec`)

Reply:
Would you prefer that I remove this from this PR or just leave it with the expectation that it's going to get changed significantly when addressing #176?

Reply:
I'm actually exploring the case where `Stack` is added in the following PR that solves #176. Like that it will be easier to see exactly what kind of implications it will have.